primitives/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,288 @@
+    Ok(F, P),
+    /// The index is valid but we do not have the leaf in memory
+    NotInMemory,
+    /// The index is outside the occupied range in the tree, and a
+    /// non-membership proof
+    EmptyLeaf(N),
+}
+
+impl<F, P, N> LookupResult<F, P, N> {
+    /// Assert the lookup result is Ok. Return a tuple of element and membership
+    /// proof.
+    pub fn expect_ok(self) -> Result<(F, P), PrimitivesError> {
+        match self {
+            LookupResult::Ok(x, proof) => Ok((x, proof)),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected Ok, found NotInMemory".to_string(),
+            )),
+            LookupResult::EmptyLeaf(_) => Err(PrimitivesError::InternalError(
+                "Expected Ok, found EmptyLeaf".to_string(),
+            )),
+        }
+    }
+
+    /// Assert the lookup result is NotInMemory. Return a non-membership proof.
+    pub fn expect_empty(self) -> Result<N, PrimitivesError> {
+        match self {
+            LookupResult::EmptyLeaf(n) => Ok(n),
+            LookupResult::Ok(..) => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found element".to_string(),
+            )),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found NotInMemory".to_string(),
+            )),
+        }
+    }
+}
+
+/// An element of a Merkle tree.
+pub trait Element: Clone + Eq + PartialEq {}
+impl<T: Clone + Eq + PartialEq> Element for T {}
+
+/// An index type of a leaf in a Merkle tree.
+pub trait Index: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath {}
+impl<T: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath> Index for T {}
+
+/// An internal node value type in a Merkle tree.
+pub trait NodeValue: Default + Eq + PartialEq + Copy + Clone + Debug {}
+impl<T: Default + Eq + PartialEq + Copy + Clone + Debug> NodeValue for T {}
+
+/// Merkle tree hash function
+pub trait DigestAlgorithm<E, I, T>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+{
+    // Possible improvement: adding digest_element() and digest_index()
+    /// Digest a list of values
+    fn digest(data: &[T]) -> T;
+
+    /// Digest a leaf (an indexed element)
+    fn digest_leaf(pos: &I, elem: &E) -> T;
+}
+
+/// An trait for Merkle tree index type.
+pub trait ToTraversalPath {
+    /// Convert the given index to a vector of branch indices given tree height
+    /// and arity.
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize>;
+}
+
+impl<F: Into<BigUint> + Clone> ToTraversalPath for F {
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize> {
+        let mut pos: BigUint = <F as Into<BigUint>>::into(self.clone());
+        let mut ret = vec![];
+        for _i in 0..height {
+            ret.push((&pos % (arity as u64)).to_usize().unwrap());
+            pos /= arity as u64;
+        }
+        ret
+    }
+}
```

Comment:
This blanket impl makes it impossible to implement `ToTraversalPath` for types other than integer-like types. For example, if it weren't for this, we could implement `ToTraversalPath` for `Vec<usize>` (trivially) or arbitrary user-defined types by serializing. I think it would be slightly better to manually implement this for each specific number type (e.g. `u8` through `u128`, `BigUint`, `U256`, field elements). It is probably not too bad to just turn this impl into a macro that gets applied for each numeric type.

Reply:
This is especially important for key-value type accumulators

Reply:
Had the same trouble before. Let me change it to a macro first.